File: docs/palindromes-in-peg/gs_dual_flags.py
"""GS overlaps of u and reverse(u), without the separator mirror word.

Pattern and text heads have the same coordinate interval [0, len(u)] but
different read views. Copies preserve the view; moves change only position.
Proper overlaps are exactly palindrome prefixes shorter than u. This saves
the doubled preprocessing length of u # reverse(u).
"""
from gs_heads import HEADS, TESTS, HeadVM, border_controller, compile_controller, unit_moves
from gs_flag_heads import FLAG_BLIND


DUAL_HEADS = (*HEADS, "TextOrigin", "Lower", "Upper", "Cursor")


def dual_flag_controller(k=8):
  yield ("copy", "Cursor", "Upper")
  yield ("move", (("Cursor", -1),))
  if not (yield ("less", "Cursor", "Lower")):
    yield from border_controller(k, flags=True, tail_origin="TextOrigin")


def compile_dual_flags(k=8, *, unit=True):
  program = compile_controller(k, controller=dual_flag_controller)
  return unit_moves(program) if unit else program


class DualFlagVM(HeadVM):
  def __init__(self, word, lower, upper, program=None):
    if not 0 <= lower <= upper <= len(word):
      raise ValueError("the proper-prefix interval must satisfy 0 <= lower <= upper <= len(u)")
    super().__init__(word, compile_dual_flags() if program is None else program)
    self.positions.update(TextOrigin=0, Lower=lower, Upper=upper, Cursor=0)
    self.reverse = dict.fromkeys(DUAL_HEADS, False)
    self.reverse["OriginalEnd"] = self.reverse["TextOrigin"] = True
    self.flags = []

  def step(self):
    event, targets = self.program.code[self.state]
    op, *args = event
    if op == "symbols":
      values = []
      for head in args:
        position = self.positions[head]
        if head in FLAG_BLIND or not 0 <= position < len(self.word):
          raise AssertionError(("invalid oriented-head read", head, position, len(self.word)))
        values.append(self.word[len(self.word) - 1 - position if self.reverse[head] else position])
      self.comparisons += 1
      self.steps += 1
      self.state = targets[int(values[0] == values[1])]
    elif op == "copy":
      target, source = args
      self.positions[target] = self.positions[source]
      self.reverse[target] = self.reverse[source]
      self.steps += 1
      self.state = targets[0]
    elif op == "move":
      for head, delta in args[0]:
        self.positions[head] += delta
        self.motion += abs(delta)
        if head not in FLAG_BLIND and not 0 <= self.positions[head] <= len(self.word):
          raise AssertionError(("oriented head left its view", head, self.positions[head]))
      self.steps += 1
      self.state = targets[0]
    elif op == "flag":
      self.flags.append(args[0])
      self.steps += 1
      self.state = targets[0]
    else:
      super().step()

  def run(self, watchdog=None):
    super().run(watchdog)
    return tuple(self.flags)

File: scala/pal/src/main/scala/pal/GsDualFlags.scala
package pal

import scala.collection.mutable

/** GS overlaps of u and reverse(u), without the separator mirror word.
  *
  * Pattern and text heads have the same coordinate interval [0, len(u)] but
  * different read views. Copies preserve the view; moves change only position.
  * Proper overlaps are exactly palindrome prefixes shorter than u. This saves
  * the doubled preprocessing length of u # reverse(u).
  *
  * Port of `gs_dual_flags.py`; see "Two oriented views" in `GS_LOCAL_CLOCK.md`.
  */
object GsDualFlags {
  import GsHeads.{HEADS, compileController, unitMoves}

  val DUAL_HEADS: Vector[String] = HEADS ++ Vector("TextOrigin", "Lower", "Upper", "Cursor")

  /** `dual_flag_controller(k)`: like `flag_controller`, with the text starting
    * at `TextOrigin` (the reverse view).
    */
  final class DualFlagController(k: Int) extends GsFlagHeads.FlagController(k, "dual_flag_controller", "TextOrigin")

  def dualFlagController(k: Int = 8): DualFlagController = new DualFlagController(k)

  def compileDualFlags(k: Int = 8, unit: Boolean = true): Program = {
    val program = compileController(k, controller = k => dualFlagController(k))
    if (unit) unitMoves(program) else program
  }
}

/** Observer for the two-view flag worker: heads carry an orientation besides a
  * coordinate; reverse-oriented heads read `u` from its end.
  */
class DualFlagVM(word: String, lower: Int, upper: Int, program: Option[Program] = None)
  extends HeadVM[Char](word.toIndexedSeq, program.getOrElse(GsDualFlags.compileDualFlags())) {
  import Event.*
  import GsDualFlags.DUAL_HEADS
  import GsFlagHeads.FLAG_BLIND

  if (!(0 <= lower && lower <= upper && upper <= word.length)) {
    throw new IllegalArgumentException("the proper-prefix interval must satisfy 0 <= lower <= upper <= len(u)")
  }
  positions("TextOrigin") = 0
  positions("Lower") = lower
  positions("Upper") = upper
  positions("Cursor") = 0

  /** Orientation of each head: `true` reads the reversed view. */
  val reverse: mutable.LinkedHashMap[String, Boolean] = mutable.LinkedHashMap.from(DUAL_HEADS.map(_ -> false))
  reverse("OriginalEnd") = true
  reverse("TextOrigin") = true

  protected val flagBuffer: mutable.ArrayBuffer[Boolean] = mutable.ArrayBuffer.empty

  /** The flag bits emitted so far, longest length first. */
  def flags: Vector[Boolean] = flagBuffer.toVector

  private def read(head: String): Char = {
    val position = positions(head)
    if (FLAG_BLIND.contains(head) || !(0 <= position && position < word.length)) {
      throw new AssertionError(("invalid oriented-head read", head, position, word.length).toString)
    }
    if (reverse(head)) word(word.length - 1 - position) else word(position)
  }

  override def step(): Unit = {
    val Row(event, targets) = current
    event match {
      case Symbols(left, right) =>
        val same = read(left) == read(right)
        comparisons += 1
        steps += 1
        state = targets(if (same) 1 else 0)
      case Copy(target, source) =>
        positions(target) = positions(source)
        reverse(target) = reverse(source)
        steps += 1
        state = targets(0)
      case Move(moves) =>
        moves.foreach { case Movement(head, delta) =>
          positions(head) += delta
          motion += math.abs(delta)
          if (!FLAG_BLIND.contains(head) && !(0 <= positions(head) && positions(head) <= word.length)) {
            throw new AssertionError(("oriented head left its view", head, positions(head)).toString)
          }
        }
        steps += 1
        state = targets(0)
      case Flag(value) =>
        flagBuffer += value
        steps += 1
        state = targets(0)
      case _ => super.step()
    }
  }

  /** Run to `halt` and return the flags (Python's `DualFlagVM.run`). */
  def runFlags(watchdog: Option[Long] = None): Vector[Boolean] = {
    run(watchdog)
    flags
  }
}
